docs: clarify the goals behind our release process (#19414)
DataDog/dd-trace-py: docs/contributing-release.rst

Reviewer: General comment: I think what you wrote here matches my understanding (and vision) of what an ideal state would be. 

Unfortunately, at the moment and for reasons laid out in my [_Refreshing the ddtrace dogweb upgrade policy_](https://docs.google.com/document/d/1tuUuE0qjx8zl4WTdoE9m3mf8VK8OHHn3CqQy6HsDZS4/edit?tab=t.0#heading=h.51cehbvy11d), I don't think we can "formalise" this process as it is not achievable with the means we have and the constraints (mainly timing-related) around releases. Those reasons also are those that led to refreshing the upgrade policy.  

The policy change has made things better, based on the last two upgrades in dogweb, but those same two last upgrades showed us that blocking RCs on dogweb would push back RCs by a lot of time -- too much time -- which isn't acceptable for product reasons (and if my understanding on that is wrong, by all means please correct me, because I'd love it to be).  

That said, I think the current version of your changes is near 100% true (realistic and even enforced today) if you add _in `dd-source` only_ in the fine print.

Author: Clarified dogweb vs ddsource

Release Philosophy
==================
Our customers run dd-trace-py in their production environments.
They rely on the stability, correctness, and performance of every release we make.
The reason customers install dd-trace-py is that it offers them features that help them solve critical business problems.

> The faster and more reliability we can deliver and iterate on these features, the more value we can provide.

Reviewer: ```suggestion
The faster and more reliably we can deliver and iterate on these features, the more value we can provide.
```

Reviewer 2: reliability is definitely something nice to deliver 😅

These dual responsibilities to our customers drive three goals of our release process, described in priority order.

1. **A stable, correct, and performant release.**
   We fully validate every release before making it available to customers.
   For a new minor release, this means that all automated tests pass, long-running integration tests have successfully completed, and we have successfully deployed the release candidate internally without incident.
   This is our top priority.

>    If a release cannot be validated, we WILL roll-back or disable new features until we find a set of features that can successfully pass all validation, including internal dogfooding.

Reviewer: As much as I'd like it to be the case, I don't think we can say that in the release docs at the moment. 

 > long-running integration tests have successfully completed

With reliability environment being in the process of becoming CuSim, I don't think we have any long-running integration tests right now 

> successfully deployed the release candidate internally without incident

This is not enforced at the moment (due to the reasons we know)

> If a release cannot be validated, we WILL roll-back or disable new features until we find a set of features that can successfully pass all validation, including internal dogfooding.

That doesn't match what we do either. It may be what we should do (or would like to do) but the release process we're actually following at the moment is more focused on unblocking RCs and releases by fixing problems than it is on reverting their root causes.  
This is made even more true by the fact dogfooding lags behind in certain codebases in which case reverting a feature once a bug is discovered is not doable (because we're past the RC step, so the new feature has already been released in a proper minor).

>    The release manager MAY declare a code-freeze speed the process of stabilizing a release.

Reviewer: ```suggestion
   The release manager MAY declare a code-freeze to speed the process of stabilizing a release.
```

2. **Timely release to enable rapid customer-centric iteration.**
   We aim to release on a predictable, frequent cadence.

>    Given the time necessary to validate a release including full dogfooding, we currently target a 2 week release cadence.

Reviewer: I don't think that's realistic 😅  getting a new version of ddtrace available and deployed in monorepos easily takes 7+ days, so two weeks is the absolute best case scenario, assuming a new release doesn't surface any bugs at all (and we most often need at least one).

We are currently aiming for a that two week release cadence but that's assuming we don't fully dogfood.

   We prefer to delay the release of a particular feature to the next release rather than delaying the entire release to fix one broken feature.

>    If a change does not make the current release train, there is another train just around the corner.

Reviewer: I'd be curious to know whether other teams agree with this. 

The standard practice today seems to be sending a "when is the next version cut? I need to add a last minute fix please" on Slack, reverting is not the default at all.

On top of this, due to the two week release cadence being a best case scenario, I don't think contributors can generally consider there's _another train just around the corner_ -- in many cases, waiting the other train means waiting another two-three weeks. My understanding is that for certain products this is not acceptable.

> 3. **Releasing customer-ready improvements as quickly as possible.**

Reviewer: What is the difference between 2 and 3?

>    A release will include as many customer improvements as are possible, given the validation and timely release requirements.

Reviewer: ```suggestion
   A release will include as many customer improvements as possible, given the validation and timely release requirements.
```

   When a change cannot be validated in time, it should be disabled or rolled-back until a subsequent release.

How to handle regressions
-------------------------

1. **Revert the change(s) that contributed the most to the regression.**
   This should be considered if the regression is not acceptable, but the fix will take longer than a day to merge to the release branch.
2. **Merge a fix to resolve the  regression.**

>    If the author(s) for the change(s) that introduced significant regression(s) consider the change to be of high priority, and are willing and able to commit to merging a fix within one day, the release manager can consider merging the change and restarting the RC process.

Reviewer: This is a dangerous line to walk, _commit to merging a fix within one day_ means the person is able to predict whether the fix will completely fix the problem (which often involves releasing services in dogfooding, etc.) 

So far what we tend to do is merge a fix if reverting is (or sounds) harder than reverting the change (e.g. because the change has since then been interleaved with other changes, such that the revert wouldn't be clean and getting confidence on it would be hard).  
But that doesn't guarantee the one day thing.

Author: clarified

If ever you discover that one of these guarantees is not upheld, please open a pull request adjusting the version string accordingly.

> 5. Declare an internal Sev5 incident to track release progress.

Reviewer 3: External contributors won't be able to follow this so I think this statement needs to be "optional", unless the alternative is we hope people open Github issues?

Reviewer: I wouldn't worry too much about that -- there are many steps of the actual release process external contributors can't do anyway, so this is just another one of them.

> 5. Declare an internal Sev5 incident to track release progress.

Reviewer: ```suggestion
5. Declare an internal SEV-5 incident to track release progress.
```

If any SLO is breached, the release pipeline on GitLab will be blocked.
See our thresholds file(s) at `bp-runner.macrobenchmarks.fail-on-breach.yml <https://github.com/DataDog/dd-trace-py/blob/3cf3342a005c1ef9e345d2a82a631bc827c8617a/.gitlab/benchmarks/bp-runner.macrobenchmarks.fail-on-breach.yml>`_ and `bp-runner.microbenchmarks.fail-on-breach.yml <https://github.com/DataDog/dd-trace-py/blob/3cf3342a005c1ef9e345d2a82a631bc827c8617a/.gitlab/benchmarks/bp-runner.microbenchmarks.fail-on-breach.yml>`_.

**If release gates fail**

Find the change(s) that contributed the most to performance regression.
You can check from the `Benchmarking Platform - Benchmarks tab <https://benchmarking.us1.prod.dog/benchmarks?projectId=3&ciJobDateStart=1753290587498&ciJobDateEnd=1753895387498&gitBranch=main>`_ and filter by project and branch to see these commits.
Notify the authors in `#apm-python-release <https://dd.enterprise.slack.com/archives/C04MK6NNDG9>`_ to see if there are any easy fixes (less than a day of work) that can be pushed to the release branch.

> Follow the revert/fix/allow policy dercribed above to remediate the issue.

Reviewer: ```suggestion
Follow the revert/fix/allow policy described above to remediate the issue.
```

See `Performance quality gates - User Guide <https://datadoghq.atlassian.net/wiki/spaces/APMINT/pages/5158175217/Performance+quality+gates+-+User+Guide>`_ for more details.

Pre-Release dogfooding
----------------------

> Following the example of dd-trace-go, all release candidates must be successfully deployed

Reviewer: I believe the latest on this is that we're _not_ following the example of `dd-trace-go` due to the significant delays in releasing (weeks) that it can incur ([link](https://dd.slack.com/archives/C04MK6NNDG9/p1785157597230959))

Still up for discussion based on that same Slack thread, but I don't expect it to happen to be honest in the very near future.

> 1. In **staging** for **24 hours**.

Reviewer: Agree with that -- staging often doesn't give signals at all, so we can keep it short.

> 2. In **prod** for **three days**.

Reviewer: This is probably enough, but it should be _three days of the service running with the latest version_ (and for all services) which tends to be hard to achieve (cf. my work in the latest upgrade incident to ping everyone to deploy their services).

Ideally we'd need a better way to do this, currently it's one of the things that drags us. Maybe it's something Python Tools could help us with (maybe for next quarter cross-team OKRs).

Author: Clarified the distinction between dogweb and ddsource

>    This means all relevant internal services (Rapid, Dogweb, McNulty etc) have successfully deployed and run the new RC without incident for that time.

Reviewer: ```suggestion
   This means all relevant internal services (both dd-source and dogweb services) have successfully deployed and run the new RC without incident for that time.
```

This is what we'd want ideally

>    Given the nature of services internal deployment pipelines, etc, there may be some services where it is infeasible to upgrade during the release window.

Reviewer: Could you clarify this?

Author: clarified
